core/matrix/sellp.hpp

```diff
@@ -0,0 +1,324 @@
+
+#ifndef GKO_CORE_MATRIX_SELLP_HPP_
+#define GKO_CORE_MATRIX_SELLP_HPP_
+
+
+#include "core/base/array.hpp"
+#include "core/base/lin_op.hpp"
+
+
+constexpr int default_slice_size = 64;
+constexpr int default_stride_factor = 1;
+
+
+namespace gko {
+namespace matrix {
+
+template <typename ValueType>
+class Dense;
+
+/**
+ * SELL-P is a matrix format similar to ELL format. The difference is that
+ * SELL-P format divides rows into smaller slices and store each slice with ELL
+ * format.
+ *
+ * @tparam ValueType  precision of matrix elements
+ * @tparam IndexType  precision of matrix indexes
+ *
+ */
+template <typename ValueType = default_precision, typename IndexType = int32>
+class Sellp : public EnableLinOp<Sellp<ValueType, IndexType>>,
+              public EnableCreateMethod<Sellp<ValueType, IndexType>>,
+              public ConvertibleTo<Dense<ValueType>>,
+              public ReadableFromMatrixData<ValueType, IndexType>,
+              public WritableToMatrixData<ValueType, IndexType> {
+    friend class EnableCreateMethod<Sellp>;
+    friend class EnablePolymorphicObject<Sellp, LinOp>;
+    friend class Dense<ValueType>;
+
+public:
+    using EnableLinOp<Sellp>::convert_to;
+    using EnableLinOp<Sellp>::move_to;
+
+    using value_type = ValueType;
+    using index_type = IndexType;
+    using mat_data = matrix_data<ValueType, IndexType>;
+
+    void convert_to(Dense<ValueType> *other) const override;
+
+    void move_to(Dense<ValueType> *other) override;
+
+    void read(const mat_data &data) override;
+
+    void write(mat_data &data) const override;
+
+    /**
+     * Returns the values of the matrix.
+     *
+     * @return the values of the matrix.
+     */
+    value_type *get_values() noexcept { return values_.get_data(); }
+
+    /**
+     * @copydoc Sellp::get_values()
+     *
+     * @note This is the constant version of the function, which can be
+     *       significantly more memory efficient than the non-constant version,
+     *       so always prefer this version.
+     */
+    const value_type *get_const_values() const noexcept
+    {
+        return values_.get_const_data();
+    }
+
+    /**
+     * Returns the column indexes of the matrix.
+     *
+     * @return the column indexes of the matrix.
+     */
+    index_type *get_col_idxs() noexcept { return col_idxs_.get_data(); }
+
+    /**
+     * @copydoc Sellp::get_col_idxs()
+     *
+     * @note This is the constant version of the function, which can be
+     *       significantly more memory efficient than the non-constant version,
+     *       so always prefer this version.
+     */
+    const index_type *get_const_col_idxs() const noexcept
+    {
+        return col_idxs_.get_const_data();
+    }
+
+    /**
+     * Returns the lengths(columns) of slices.
+     *
+     * @return the lengths(columns) of slices.
+     */
+    size_type *get_slice_lens() noexcept { return slice_lens_.get_data(); }
```

Comment:
why do you call this "lens" ? Do you mean "length"?

Reply:
@hartwiganzt 
Yes, and the "length" specifically means the number of `max_nonzero_per_row` in each slice.

Reply:
then lets better use "get_slice_lengths"